cpp/ql/src/semmle/code/cpp/models/implementations/Gets.qll

```diff
@@ -0,0 +1,43 @@
+import semmle.code.cpp.models.interfaces.DataFlow
+import semmle.code.cpp.models.interfaces.Taint
+import semmle.code.cpp.models.interfaces.ArrayFunction
+import semmle.code.cpp.models.interfaces.Alias
+import semmle.code.cpp.models.interfaces.SideEffect
+
+/**
+ * The standard functions `gets` and `fgets`.
+ */
+class GetsFunction extends DataFlowFunction, TaintFunction, ArrayFunction, AliasFunction, SideEffectFunction {
+  GetsFunction() {
+    exists(string name | name = getName() |
+      name = "gets" or // gets(str)
+      name = "fgets" // fgets(str, num, stream)
+    )
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isReturnValue()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameterDeref(2) and
```

Comment:
If a `FILE *fh` is tainted, presumably via a query-specific source, should we think of that taint as being on `fh` or `*fh`? I've always thought of `FILE *` as an opaque handle, and I was surprised to learn it's possible to dereference it, at least with glibc on Linux. According to https://en.cppreference.com/w/c/io, a C library doesn't have to allow `FILE *` to be dereferenced, and it's undefined behaviour to create your own `FILE *` values. Because of all that, I think `isParameter(2)` would be more appropriate here. It's also more practical as long as we don't have a "return deref write side effect" instruction.

```suggestion
    input.isParameter(2) and
```

Reply:
Yep, I'm convinced. :+1: